### ISSUE TITLE
Consider the use of `unique` for identical symmetries in `Orientation.dot`

### ISSUE THREAD

**Issue by harripj** [pyxem/orix#296]
Consider the case where one wishes to find the angle between two `Orientation` with the same symmetry, say `Oh`. The following calculation is currently performed: https://github.com/pyxem/orix/blob/d474f246fedd1fabc44f65f103db7ee096759a53/orix/quaternion/orientation.py#L570-L575

In my testing on my machine the total time for this operation is `~5.5ms`, however around `5ms` of this is the `unique` operation. Perhaps we introduce a test where the `unique` function is only performed if the symmetries are different.

Further to this, if we compute the maximum angle after the symmetry unique operation, we find that nothing changes:
```python
from orix.quaternion import symmetry

for name in dir(symmetry):
    s = getattr(symmetry, name)
    if isinstance(s, symmetry.Symmetry):
        u = s.outer(s).unique()
        if s.size != u.size:
            print(name, u, s)
        else:
            delta = (s * ~u).angle.data
            print(name, np.allclose(delta, 0))
```
```
C1 True
C2 True
C2h True
C2v True
C2x True
C2y True
C2z True
C3 True
C3h True
C3v True
C3x True
C3y True
C3z True
C4 True
C4h True
C4v True
C4x True
C4y True
C4z True
C6 True
C6h True
C6v True
Ci True
Cs True
Csx True
Csy True
Csz True
D2 True
D2d True
D2h True
D3 True
D3d True
D3h True
D3x True
D3y True
D4 True
D4h True
D6 True
D6h True
O True
Oh True
S4 True
S6 True
T True
Td True
Th True
_cubic Symmetry (3,) 
[[ 1.   0.   0.   0. ]
 [ 0.5  0.5  0.5  0.5]
 [-0.5  0.5  0.5  0.5]] Symmetry (2,) 
[[1.  0.  0.  0. ]
 [0.5 0.5 0.5 0.5]]
_mirror_xy True
```

which seems to be unnecessary computation apart from `_cubic` symmetry.

What I would suggest then is to add a check to see whether to the two symmetries are the same, and if they are, avoid the unique computation.

On this note, there is an `angle_with` function and I think it would be a good idea to include an `angle_with_outer` function too.

**Comment by harripj** [comment]
As some symmetries are related, perhaps a different test would be:
```python
s1, s2 = symmetry.Oh, symmetry.Th

axis = np.array((s1.size, s2.size)).argmax()
angles = s1.outer(s2).angle.data.min(axis=axis)
if np.allclose(angles, 0)...
```
 Time comparison:
```python
>>> %timeit s1.outer(s2).angle.data.min(axis=axis)
299 µs ± 34.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> %timeit s1.outer(s2).unique()
2.6 ms ± 445 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by hakonanes** [comment]
I wrote the `Orientation.dot()` method (commit https://github.com/pyxem/orix/commit/c312338786e9782d9f637ff66898d4c78e3e9a22) originally to add symmetry reduced `angle_with()` for `Orientation` instances, which previously used `Rotation.angle_with()` which didn't consider symmetry. I based the implementation on [`Misorientation._distance()`](https://github.com/pyxem/orix/blob/master/orix/quaternion/orientation.py#L85) and, more importantly, on MTEX' `orientation.dot`, where they reduce the symmetry operations to the unique ones:

https://github.com/mtex-toolbox/mtex/blob/6c995ce5531160a6a96f87cb68a968e5166330c4/geometry/%40orientation/dot.m#L43-L60

I thought perhaps we reduced the memory footprint with the call to `unique()`, but your tests show that we don't? I'm all for adding some checks where appropriate to reduce computation time. I cannot say immediately what they should be, though. Before adding these checks to `angle_with()`, we should expand [our tests](https://github.com/pyxem/orix/blob/master/orix/tests/quaternion/test_orientation.py#L396) to all point groups and several random orientations to ensure that angles are correct, preferably compared to MTEX.

An `Orientation.angle_with_outer()` sounds good. [`Orientation.get_distance_matrix()`](https://orix.readthedocs.io/en/stable/reference.html#orix.quaternion.Orientation.get_distance_matrix) is a special case where the outer angles with itself are compuated. Computation is done with Dask because it is computationally memory intensive. This method should be re-considered with an `angle_with_outer()`, to keep memory in check and don't repeat too much code.

**Comment by harripj** [comment]
> I wrote the Orientation.dot() method (commit https://github.com/pyxem/orix/commit/c312338786e9782d9f637ff66898d4c78e3e9a22) originally to add symmetry reduced angle_with() for Orientation instances, which previously used Rotation.angle_with() which didn't consider symmetry. I based the implementation on [Misorientation._distance()](https://github.com/pyxem/orix/blob/master/orix/quaternion/orientation.py#L85) and, more importantly, on MTEX' [orientation.dot](https://github.com/mtex-toolbox/mtex/blob/6c995ce5531160a6a96f87cb68a968e5166330c4/geometry/%40orientation/dot.m#L43-L60), where they reduce the symmetry operations to the unique ones.

Thanks for adding some clarity, I agree that reducing the symmetry orientations to only the unique ones, and also taking inspiration from MTEX are both good ways forward!

> I thought perhaps we reduced the memory footprint with the call to unique(), but your tests show that we don't?

I haven't checked the memory footprint, and I am only talking about the specific case where the `symmetry` of `self` and `other` are identical (or perhaps related to one another, eg. `Oh` and `Th` in the second example. For `Oh` with `Oh` for example, the call to unique takes 90% of the computational time of the function and I think can be avoided as the two symmetries are identical, so the unique outer products are the same as the symmetry itself. This is what I was trying to show in the first example. For this case a simple check of `if self.symmetry is other.symmetry` would be sufficient to skip the unique call and significantly reduce overall computational time. If the check evaluates `False` then of course we keep the flow of the current working code.

> Before adding these checks to angle_with(), we should expand [our tests](https://github.com/pyxem/orix/blob/master/orix/tests/quaternion/test_orientation.py#L396) to all point groups and several random orientations to ensure that angles are correct, preferably compared to MTEX.

I agree and this is very important.

> An Orientation.angle_with_outer() sounds good. [Orientation.get_distance_matrix()](https://orix.readthedocs.io/en/stable/reference.html#orix.quaternion.Orientation.get_distance_matrix) is a special case where the outer angles with itself are compuated. Computation is done with Dask because it is computationally memory intensive. This method should be re-considered with an angle_with_outer(), to keep memory in check and don't repeat too much code.

This sounds like it could be easily implemented as it is essentially already written! 😄

**Comment by hakonanes** [comment]
> For this case a simple check of if self.symmetry is other.symmetry would be sufficient to skip the unique call and significantly reduce overall computational time.

I suggest we make this check and skip the call to unique in this case. Seems like low hanging fruit. More complicated tests can be introduced, but should be more thoroughly tested (opened #300 with this in mind).

**Comment by pc494** [comment]
closed by #303
